text/0000-epochs.md

```diff
@@ -0,0 +1,533 @@
+
+There are a few reasons to provide such a preview:
+
+- Most importantly, it clears the way to shipping features for the next epoch on
+  the stable channel as they become ready, even if they require some existing
+  usages to become errors. Again, keyword introduction is a simple example: the
+  `2015-next` epoch can begin by making it a hard error to use `catch` as an
+  identifier (which is allowed today), and then later stabilizing the new
+  `catch` feature when it is ready. By "locking in" the hard errors up front,
+  however, **the preview of the epoch is guaranteed to be stable**: once code
+  compiles on the preview epoch, it will continue to do so as further features
+  stabilize.
+
+- The preview epoch also gives a clear picture of exactly what deprecations will
+  become errors in the next epoch, providing a nice way to future-proof your
+  code for upgrading to the epoch when available.
+
+Putting this all together, in an "epoch year" the roadmap will lay out an
+expected set of deprecations-to-become-errors and potential features; early in
+the year, an epoch preview will be released which makes all the slated
+deprecations into errors. As the year progresses, features will continue
+stabilize but some may only be available by opting into the preview epoch (since
+they rely on e.g. new keywords). Finally, when we're ready to ship the full new
+product, we enable the new epoch and deprecate the preview version (which will
+just be an alias for it).
+
+There are some alternative ways to achieve similar ends, but with significant
+downsides; these are explored in the Alternatives section.
+
+## Constraints on epoch changes
+
+We have already constrained epochal changes to essentially removing deprecations
+(and thus making way for new features). But we want to further constrain them,
+for two reasons:
+
+- **Limiting technical debt**. The compiler retains compatibility for old
+  epochs, and thus must have distinct "modes" for dealing with them. We need to
+  strongly limit the amount and complexity of code needed for these modes, or
+  the compiler will become very difficult to maintain.
+
+- **Limiting deep conceptual changes**. Just as we want to keep the compiler
+  maintainable, so too do we want to keep the conceptual model sustainable. That
+  is, if we make truly radical changes in a new epoch, it will be very difficult
+  for people to reason about code involving different epochs, or to remember the
+  precise differences.
+
+As such, the RFC proposes to limit epochal changes to be "superficial",
+i.e. occurring purely in the early front-end stages of the compiler. More
+concretely:
+
+- We identify "core Rust" as being, roughly, MIR and the core trait system.
+  - Over time, we'll want to make this definition more precise, but this is best
+    done in an iterative fashion rather than specified completely up front.
+- Epochs can only change the front-end translation into core Rust.
```

Comment:
So, what would you make of changes like const fn? To the eye of someone totally oblivious about compiler design like me, they seem to affect more than the translation of _run-time_ code...